lib/cli/src/Cardano/CLI.hs

```diff
@@ -388,43 +388,83 @@ parseSeedWords = some (argument str (metavar "MNEMONIC_WORDS..."))
 
 -- | Lay thine eyes upon the type parameters and see that there are none.
 data CliKeyScheme = CliKeyScheme
-    { wordsToSeed :: [Text] -> Either String (Passphrase "seed")
-    , allowedWordLengths :: [Int]
-    , seedToRootKey :: Passphrase "seed" -> XPrv
+    { allowedWordLengths :: [Int]
+    , mnemonicToRootKey :: [Text] -> Either String XPrv
     }
 
 byronCliKeyScheme :: ByronWalletStyle -> CliKeyScheme
 byronCliKeyScheme = \case
-    Random -> mkScheme (Proxy @'Random)
-            (Byron.getKey . flip Byron.generateKeyFromSeed pass)
+    Random ->
+        let
+            proxy = (Proxy @'Random)
+        in
+            CliKeyScheme
+                (allowedLengths proxy)
+                (fmap byronKeyFromSeed . seedFromMnemonic proxy)
     Icarus ->
-        mkScheme (Proxy @'Icarus)
-            icarusStyleKeyToSeed
+        let
+            proxy = (Proxy @'Icarus)
+        in
+            CliKeyScheme
+                (allowedLengths proxy)
+                (fmap icarusKeyFromSeed . seedFromMnemonic proxy)
     Trezor ->
-        mkScheme (Proxy @'Trezor)
-            icarusStyleKeyToSeed
+        let
+            proxy = (Proxy @'Trezor)
+        in
+            CliKeyScheme
+                (allowedLengths proxy)
+                (hardwareKeyFromMnemonic)
+
     Ledger ->
-        mkScheme (Proxy @'Ledger)
-            icarusStyleKeyToSeed
+        let
+            proxy = (Proxy @'Ledger)
+        in
+            CliKeyScheme
+                (allowedLengths proxy)
+                hardwareKeyFromMnemonic
+
   where
-    mkScheme
+    seedFromMnemonic
         :: forall (s :: ByronWalletStyle).
-            ( FromMnemonic (AllowedMnemonics s) "seed"
-            , ManyNatVal (AllowedMnemonics s)
-            )
+            (FromMnemonic (AllowedMnemonics s) "seed")
         => Proxy s
-        -> (Passphrase "seed" -> XPrv)
-        -> CliKeyScheme
+        -> [Text]
+        -> Either String (Passphrase "seed")
+    seedFromMnemonic _ =
+        left getFromMnemonicError . fromMnemonic @(AllowedMnemonics s)
 
-    -- Construct a KeyScheme using the type families of the promoted type
-    -- argument.
-    mkScheme _ = CliKeyScheme
-        (left getFromMnemonicError . fromMnemonic @(AllowedMnemonics s) @"seed")
-        (map fromIntegral (manyNatVal $ Proxy @(AllowedMnemonics s)))
+    allowedLengths
+        :: forall (s :: ByronWalletStyle). ( ManyNatVal (AllowedMnemonics s))
+        => Proxy s
+        -> [Int]
+    allowedLengths _ =
+         (map fromIntegral (manyNatVal $ Proxy @(AllowedMnemonics s)))
 
-    icarusStyleKeyToSeed =
+    icarusKeyFromSeed =
         (Icarus.getKey . flip Icarus.generateKeyFromSeed pass)
 
+    byronKeyFromSeed = (Byron.getKey . flip Byron.generateKeyFromSeed pass)
+
+    hardwareKeyFromMnemonic :: [Text] -> Either String XPrv
+    hardwareKeyFromMnemonic mw = Icarus.getKey <$> case length mw of
```

Comment:
This is now very ugly 🤔 
- It is duplicated with the API.
- It does **not** respect the `AllowedMnemonics`.

Reply:
We'd need something like
https://github.com/input-output-hk/cardano-wallet/blob/09242e8b7a6d9fd367796d2c82853e49bd37de48/lib/core/src/Cardano/Wallet/Primitive/AddressDerivation.hs#L392-L401

Maybe:
- we can return some `forall mw. (Mnemonic mw)` instead of `Passphrase purpose` (in fromMnemonic)
- we can adjust some existing pieces (like `generateKeyFromHardwareLedger`) to work better together somehow
- we can remove the `mw` type parameter on `Mnemonic mw`. I've grown more and more questioning of it at least.

Reply:
Resolved with #1343